### PR TITLE
4311 fix search results not shown

### LINKED ISSUE #4311: Search results do not appear when time frame is increased
**Description**
When you enter a search term and then increase the time frame in the search view, new results (if available) aren't displayed until the search term is deleted and entered again.

**To Reproduce**
1. Type in a search query
2. Click on "Show more"
3. Increase the time frame
4. Accept indexing of more emails
5. New results do not show up
6. Delete search term and enter it again
7. See results

**Expected behavior**
New search results should be automatically shown after indexing is done.

**Test notes**

- [ ] follow the reproduction steps and verify expected behavior 
- regression test search:
- [ ]  trigger search from search bar
- [ ] trigger search by changing the url (e.g. going back)
- [ ] trigger search from search view:
   - [ ] increase and decrease time period
   - [ ] search for different fields
   - [ ] search for different folders
   - [ ] scroll down to load more emails
- [ ]  test searching while indexing is still in progress

### PR DESCRIPTION
**Pull request by sarashub**
close #4311

### CODE REVIEW

File: src/api/worker/search/MailIndexer.ts [hunk 1]
```diff
@@ -225,23 +225,16 @@ export class MailIndexer {
...
-					   }
-				   })
-				   .catch(
-					   ofClass(CancelledError, e => {
-						   console.log("extend mail index has been cancelled", e)
-					   }),
-				   )
+	async extendIndexIfNeeded(user: User, newOldestTimestamp: number): Promise<void> {
+		if (this.currentIndexTimestamp > FULL_INDEXED_TIMESTAMP && this.currentIndexTimestamp > newOldestTimestamp) {
+			this.mailboxIndexingPromise = this.mailboxIndexingPromise.then(async () => {
```
Comment (charlag) [inline src/api/worker/search/MailIndexer.ts:230]: we are trying to chain after the previous one but if one of them fails with an exception we won't be able to continue ever, they will all reject.
@nowhub  did something similar recently: https://github.com/tutao/tutanota/blob/6999945350c860507e2e25a5a6a8527e1b2fb9fc/src/misc/UsageTestModel.ts#L377
to cover both branches.

File: src/api/worker/search/SearchTypes.ts [hunk 2]
```diff
@@ -156,6 +155,7 @@ export type SearchIndexStateInfo = {
 	mailIndexEnabled: boolean
 	progress: number
 	currentMailIndexTimestamp: number
+	aimedMailIndexTimestamp: number // this is the timestamp we are currently indexing for (or same as currentMailIndexTimestamp if we are not indexing)
```
Comment (charlag) [inline src/api/worker/search/SearchTypes.ts:158]: I think this could be a jsdoc comment

File: src/search/SearchBar.ts [hunk 3]
```diff
@@ -301,6 +303,10 @@ export class SearchBar implements Component<SearchBarAttrs> {
 		}
 	}
 
+	timePeriodChanged(oldEnd: number | null, aimedEnd: number): boolean {
```
Comment (charlag) [inline src/search/SearchBar.ts:306]: could be private

maybe should be named `timePeriodHasChanged`

File: src/search/SearchBar.ts [hunk 4]
```diff
@@ -150,7 +150,9 @@ export class SearchBar implements Component<SearchBarAttrs> {
 										!indexState.failedIndexingUpTo &&
 										currentResult &&
 										this._state().indexState.progress !== 0 &&
-										indexState.progress === 0
+										indexState.progress === 0 &&
+										//if period is changed from search view a new search is triggered there,  and we do not want to overwrite its result
+										!this.timePeriodChanged(currentResult.restriction.end, indexState.aimedMailIndexTimestamp)
```
Comment (charlag) [inline src/search/SearchBar.ts:155]: this only prevents triggering the search from the search bar. Do we need to take care of triggering the search via URL, e.g. going back? We can also address it separately but I wanted to bring it to your attention

File: src/api/worker/search/SearchTypes.ts [hunk 5]
```diff
@@ -150,12 +150,16 @@ export const enum IndexingErrorReason {
 	ConnectionLost
 }
 
+/**
+ * Current state of the Mailindexer
+ * aimedMailIndexTimestamp is the timestamp we are currently indexing for (or same as currentMailIndexTimestamp if we are not indexing)
```
Comment (charlag) [inline src/api/worker/search/SearchTypes.ts:155]: you can also put it on a single field but it's okay here

File: src/search/SearchBar.ts (hunk 4 above)
Comment (sarashub) [reply src/search/SearchBar.ts]: updating the url only triggers a search if `SearchModel.isNewSearch` is true, so it should be fine I think.

**Review by charlag — APPROVED** [review]

**Pull request merged**